harness/determined/profiler.py

```diff
@@ -18,19 +18,96 @@
 MAX_COLLECTION_SECONDS = 300
 LOG_NAMESPACE = "determined-profiler"
 
-try:
-    import pynvml
 
-    pynvml.nvmlInit()
-    SHOULD_PROFILE_GPUS = True
-except ModuleNotFoundError:
-    logging.info(f"{LOG_NAMESPACE}: pynvml not found. Not collecting GPU metrics")
-    SHOULD_PROFILE_GPUS = False
-except pynvml.NVMLError_LibraryNotFound:
-    logging.info(f"{LOG_NAMESPACE}: pynvml LibraryNotFound error. Not collecting GPU metrics")
-    SHOULD_PROFILE_GPUS = False
-except Exception as e:
-    raise RuntimeError(f"Could not set up pynvml, but it failed with an unexpected error: {e}")
+class PynvmlWrapperError(Exception):
+    pass
+
+
+class PynvmlWrapper:
+    def __init__(self) -> None:
+        self._pynvml = None  # type: Optional[Any]
+        self._device_count = None  # type: Optional[int]
+        self._index_to_uuid_map = {}  # type: Dict[int, str]
+        try:
+            import pynvml
+
+            pynvml.nvmlInit()
+            try:
+                num_gpus = pynvml.nvmlDeviceGetCount()
+                for i in range(num_gpus):
+                    handle = pynvml.nvmlDeviceGetHandleByIndex(i)
+                    pynvml.nvmlDeviceGetUUID(handle)
+                    pynvml.nvmlDeviceGetMemoryInfo(handle)
+                    pynvml.nvmlDeviceGetUtilizationRates(handle)
+                self._pynvml = pynvml
+            except Exception as e:
+                logging.warning(
+                    f"{LOG_NAMESPACE}: pynvml is functional, but failed to pass functionality "
+                    f"test due to exception. Not collecting GPU metrics. Exception details: {e}"
+                )
+        except ModuleNotFoundError:
+            logging.info(f"{LOG_NAMESPACE}: pynvml not found. Not collecting GPU metrics")
+        except pynvml.NVMLError_LibraryNotFound:
+            logging.info(
+                f"{LOG_NAMESPACE}: pynvml LibraryNotFound error. Not collecting GPU metrics"
+            )
+        except Exception as e:
```

Comment:
Is this even possible? I think we just shouldn't even catch this one maybe? Idk, I guess this is strictly better since it tells users exactly where/what happened.